src/ecredis.erl

```diff
@@ -0,0 +1,420 @@
+    Result.
+
+
+%% Executes qp.
+-spec qp(ClusterName::atom(), redis_pipeline_command()) -> redis_pipeline_result().
+qp(ClusterName, Commands) ->
+    q(ClusterName, Commands).
+
+
+%% Executes q.
+-spec q(ClusterName::atom(), redis_command()) -> redis_result().
+q(ClusterName, Command) ->
+    query(ClusterName, Command).
+
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+% Internal methods
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+
+-spec connect_(State::#state{}, [{Address::string(), Port::integer()}]) -> #state{}.
+connect_(_State, []) ->
+    #state{};
+connect_(State, InitNodes) ->
+    NewState = State#state{
+        slots_cache = {},
+        slots_maps = {},
+        init_nodes = [#node{address = A, port = P} || {A,P} <- InitNodes],
+        version = 0
+    },
+    reload_slots_map(NewState).
+
+
+-spec remap_cluster_internal(State::#state{}, Version::integer()) -> State::#state{}.
+remap_cluster_internal(State, Version) ->
+    if State#state.version == Version ->
+        reload_slots_map(State);
+    true ->
+        State
+    end.
+
+
+%% TODO(vipin): Need to reset connection on Redis node removal.
+-spec reload_slots_map(State::#state{}) -> State::#state{}.
+reload_slots_map(State) ->
+    ClusterSlots = get_cluster_slots(State#state.cluster_name, State#state.init_nodes),
+    SlotsMaps = parse_cluster_slots(ClusterSlots),
+    ConnectedSlotsMaps = connect_all_slots(State#state.cluster_name, SlotsMaps),
+    SlotsCache = create_slots_cache(ConnectedSlotsMaps),
+
+    NewState = State#state{
+        slots_cache = list_to_tuple(SlotsCache),
+        slots_maps = list_to_tuple(ConnectedSlotsMaps),
+        version = State#state.version + 1
+    },
+    [cache_eredis_pids(NewState, Slot) || Slot <- lists:seq(0, ?REDIS_CLUSTER_HASH_SLOTS - 1)],
+
+    %% Reset tuples not needed in the state.
+    NewState2 = NewState#state{
+      slots_cache = {},
+      slots_maps = {}
+    },
+    NewState2.
+
+-spec get_cluster_slots(ClusterName::atom(), [#node{}]) -> [[bitstring() | [bitstring()]]].
+get_cluster_slots(_ClusterName, []) ->
+    throw({error,cannot_connect_to_cluster});
+get_cluster_slots(ClusterName, [Node|T]) ->
+    Res = lookup_eredis_pid(ClusterName, Node#node.address, Node#node.port),
+    case Res of
+        {ok, Pid} ->
+          case eredis:q(Pid, ["CLUSTER", "SLOTS"]) of
+            {error,<<"ERR unknown command 'CLUSTER'">>} ->
+                get_cluster_slots_from_single_node(Node);
+            {error,<<"ERR This instance has cluster support disabled">>} ->
+                get_cluster_slots_from_single_node(Node);
+            {ok, ClusterInfo} ->
+                ClusterInfo;
+            _ ->
+                get_cluster_slots(ClusterName, T)
+          end;
+        _ ->
+            get_cluster_slots(ClusterName, T)
+  end.
+
+
+-spec get_cluster_slots_from_single_node(#node{}) -> [[bitstring() | [bitstring()]]].
```

Comment:
fix spec to binary(): in all places in the file unless necessary..

Reply:
I don't understand. Can you give example?

Reply:
```suggestion
-spec get_cluster_slots_from_single_node(#node{}) -> [[binary() | [binary()]]].
```